npm-command-guidelines.md

```diff
@@ -0,0 +1,60 @@
+# NPM Command Structure
+
+##What is this?
+This document is an attempt to loosely standardise the names of npm commands we use within our JavaScript projects. This is being proposed to introduce a level of familiarity between projects so that:
+
+ - The learning time common project commands for new developers is shortened.
+ - Context switching between projects is less of a headache.
+ - Developers make fewer mistakes during project setup.
+ - Boilerplate project code becomes more reusable, shortening time to create new projects and encourage similarity between projects at the same time.
+
+This guide has been developed by analysing which npm commands and command patterns are already in use at Holiday Extras, and doing our best to preserve that while establishing a common standard.
+
+##Observations from our repositories
+The following observations have been made from analysing the scripts blocks of all of the `package.json` files within the Holiday Extras Github organisation's projects.
+
+###Types of command
+Our npm commands tend to be structured into 5 main categories:
+
+ - Build related commands
+ - Test related commands
+ - Run related commands
+ - CI related commands
+ - Deploy/Publish related commands
+
+There's also a long tail of misellaneous commands.
+
+###Command structure
+We seem to almost universally use `:` as a separator to namespace our commands. E.g. `build:js` and `build:css`.
+
+###Commands to run at a certain time
+We have a not insignificant number of commands with names that reflect _when_ they should be run, e.g. `prestart`, `pretest` and `postci`. This is in contrast to the majority of our commands have names that reflect _what_ the command does.
+
+##Recommendations
+Based on the above, we've come up with the following recommendations for command names.
+
+### Namespace commands
+Commands should be namespaced sensibly in a `:` separated format where possible. Common namespaces that we should aim to share between our projects should include `build`, `test`, `run`, `ci`, `deploy`, and `publish`.
+
+###Reserve the top level of a namespace for a 'do-all' command
+While we do want to namespace, we don't want to stop people from using the top-level `npm run test` or `npm run build` commands. These should, however, be reserved for a 'do everything' task _if_ one is readily available.
+
+For example, `npm run test` should run _all_ the tests of _all_ types, whereas you may have `npm run test:unit`, `npm run test:integration` and `npm run test:journey` if you wanted to split up your test types into individual commands.
```

Comment:
in SB we agreed to namespace `coverage` as `test:coverage`... SB scripts are here https://github.com/holidayextras/the-works/blob/staging/package.json#L20 - some optimisations still to be made I'm sure. . Some explanation:
```
    "ci": "istanbul cover _mocha -- -R xunit > $CIRCLE_TEST_REPORTS/unit/results.xml", // could be npm run test:coverage -- ...
    "postci": "istanbul check-coverage", // so runs after the ci
    "lint": "make-up lib config controllers flows plugins test",
    "prerelease": "node_modules/deployment-helpers/nodeApps/preRelease.sh",
    "start": "scripts/startup.sh", // this is non standard, because the start command is too long to go here
    "pretest": "npm run prerelease",
    "test": "npm run lint && npm run test:coverage",
    "test:coverage": "istanbul cover _mocha", // works because we have config in .istanbul.yml and mocha.opts
    "posttest": "istanbul check-coverage" // always runs after test
```
then mocha.opts has things that were passed in in the package.json
```
--timeout 15000
--recursive
--reporter spec
```
(plus also `--require test/helpers`) and .istanbul.yml has
```
instrumentation:
  include-all-sources: true
```
so we're reporting on all code